src/Compiler/SyntaxTree/SyntaxTreeOps.fs

```diff
@@ -982,13 +982,39 @@ let rec synExprContainsError inpExpr =
 
     walkExpr inpExpr
 
+let longIdentToString (ident: SynLongIdent) =
+    System.String.Join(".", ident.LongIdent |> List.map (fun ident -> ident.idText.ToString()))
+
 let (|ParsedHashDirectiveArguments|) (input: ParsedHashDirectiveArgument list) =
     List.map
         (function
         | ParsedHashDirectiveArgument.String(s, _, _) -> s
-        | ParsedHashDirectiveArgument.SourceIdentifier(_, v, _) -> v)
+        | ParsedHashDirectiveArgument.Int32(n, _) -> string n
+        | ParsedHashDirectiveArgument.SourceIdentifier(_, v, _) -> v
+        | ParsedHashDirectiveArgument.Ident(ident, _) -> ident.idText
+        | ParsedHashDirectiveArgument.LongIdent(ident, _) -> longIdentToString ident)
         input
 
+let (|ParsedHashDirectiveStringArguments|) (input: ParsedHashDirectiveArgument list) =
+
+    let value =
+        List.map
+            (function
+            | ParsedHashDirectiveArgument.String(s, _, _) -> Some s
+            | ParsedHashDirectiveArgument.Int32(n, m) ->
+                errorR (Error(FSComp.SR.featureParsedHashDirectiveUnexpectedInteger (n), m))
+                None
+            | ParsedHashDirectiveArgument.SourceIdentifier(_, v, _) -> Some v
+            | ParsedHashDirectiveArgument.Ident(ident, m) ->
+                errorR (Error(FSComp.SR.featureParsedHashDirectiveUnexpectedIdentifier (ident.idText), m))
+                None
+            | ParsedHashDirectiveArgument.LongIdent(ident, m) ->
+                errorR (Error(FSComp.SR.featureParsedHashDirectiveUnexpectedIdentifier (longIdentToString ident), m))
+                None)
+            input
+
+    value |> List.filter Option.isSome |> List.map (Option.defaultValue "")
```

Comment:
nit: `map` + `filter` on `option` ==> `choose`

This should work:

```suggestion
    input
    |> List.choose
        (function
        | ParsedHashDirectiveArgument.String(s, _, _) -> Some s
        | ParsedHashDirectiveArgument.Int32(n, m) ->
            errorR (Error(FSComp.SR.featureParsedHashDirectiveUnexpectedInteger (n), m))
            None
        | ParsedHashDirectiveArgument.SourceIdentifier(_, v, _) -> Some v
        | ParsedHashDirectiveArgument.Ident(ident, m) ->
            errorR (Error(FSComp.SR.featureParsedHashDirectiveUnexpectedIdentifier (ident.idText), m))
            None
        | ParsedHashDirectiveArgument.LongIdent(ident, m) ->
            errorR (Error(FSComp.SR.featureParsedHashDirectiveUnexpectedIdentifier (longIdentToString ident), m))
            None)
```

Reply:
done, thanks.